calendar/calendar-impl/impl/src/java/org/sakaiproject/calendar/impl/BaseExternalCalendarSubscriptionService.java

```diff
@@ -314,10 +318,36 @@ public Calendar getCalendarSubscription(String reference)
 		m_log.debug("ExternalCalendarSubscriptionService.getCalendarSubscription("
 				+ reference + ")");
 		m_log.debug(" |-> subscriptionUrl: " + subscriptionUrl);
+		
+		String currentSessionUserId = m_sessionManager.getCurrentSessionUserId();
+		
+		//External references will not work for anonymous users, no idea how to track circular references for them.
+		if ((currentSessionUserId==null)||(currentSessionUserId.isEmpty())) return null;
 
-		ExternalSubscription subscription = getExternalSubscription(subscriptionUrl,
-				_ref.getContext());
-
+		ExternalSubscription subscription = null;
+		if (!externalCalRefs.containsKey(currentSessionUserId))
+		{
+			externalCalRefs.put(currentSessionUserId, subscriptionUrl);
+			subscription = getExternalSubscription(subscriptionUrl, _ref.getContext());
+			externalCalRefs.remove(currentSessionUserId);
+		}
+		else
+		{
+			String urlList = externalCalRefs.get(currentSessionUserId);
+			if (urlList.contains(subscriptionUrl))
+			{
+				m_log.debug("Reached circular reference for external calendar: "+subscriptionUrl);
+				return null;
+			}
+			else
+			{
+				urlList = urlList + ";" + subscriptionUrl;
+				externalCalRefs.replace(currentSessionUserId, urlList);
+				subscription = getExternalSubscription(subscriptionUrl, _ref.getContext());
+				urlList= urlList.replace(subscriptionUrl, "").replaceAll(";;", ";");
+				externalCalRefs.replace(currentSessionUserId, urlList);
```

Comment:
Maybe I'm missing something here, but why are you calling `.replace()` on the map twice here? Won't the first one never be used essentially?